From 91b0db57f7b7d0a3f919d0e1ac1750f1928814d9 Mon Sep 17 00:00:00 2001
From: Lars Kanis <kanis@comcard.de>
Date: Mon, 4 Nov 2019 11:00:13 +0100
Subject: [PATCH] Fix coroutine support on win32

Ruby master branch currently fails on win32 MINGW at this spec:
https://github.com/ruby/spec/blob/master/core/thread/element_set_spec.rb

MINGW makes use of setjmp3() implemented in MSVCRT.DLL.
This function traverses the SEH list up to a terminating pointer 0xFFFFFFFF.
It therefore currently segfaults on NULL.
The SEH linked list must be terminated by 0xFFFFFFFF instead of NULL.

This fixes the issue mentioned here:
  https://github.com/ruby/ruby/pull/2279#issuecomment-509508810
---
 coroutine/win32/Context.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/coroutine/win32/Context.h b/coroutine/win32/Context.h
index c70b65cd2f..299515ed92 100644
--- a/coroutine/win32/Context.h
+++ b/coroutine/win32/Context.h
@@ -42,7 +42,7 @@ static inline void coroutine_initialize(
     *--context->stack_pointer = (void*)start;
 
     /* Windows Thread Information Block */
-    *--context->stack_pointer = 0; /* fs:[0] */
+    *--context->stack_pointer = (void*)0xFFFFFFFF; /* fs:[0] */
     *--context->stack_pointer = (void*)top; /* fs:[4] */
     *--context->stack_pointer = (void*)stack;  /* fs:[8] */
 
-- 
2.19.1.windows.1

